src/error.rs

```diff
@@ -289,22 +284,45 @@ fn parse_fields<'input, 'state>(
     })
 }
 
-fn is_type_ends_with(ty: &syn::Type, tail: &str) -> bool {
+/// Checks if `ty` is [`syn::Type::Path`] and ends with segment matching `tail`
+/// and doesn't contain any generic parameters.
+fn is_type_path_ends_with_segment(ty: &syn::Type, tail: &str) -> bool {
     let ty = match ty {
         syn::Type::Path(ty) => ty,
         _ => return false,
     };
 
     // Unwrapping is safe, cause 'syn::TypePath.path.segments'
     // have to have at least one segment
-    let ty = ty.path.segments.last().unwrap();
+    let segment = ty.path.segments.last().unwrap();
 
-    match ty.arguments {
+    match segment.arguments {
         syn::PathArguments::None => (),
         _ => return false,
     };
 
-    ty.ident == tail
+    segment.ident == tail
+}
+
+fn infer_source_field(state: &State, parsed_fields: &ParsedFields) -> Option<usize> {
+    let len_source_backtrace = (
+        state.fields.len(),
+        parsed_fields.source,
+        parsed_fields.backtrace,
+    );
+
+    // If we have exactly two fields, one of the fields was specified/inferred as backtrace field,
+    // but no source field was specified/inferred...
+    if let (2, None, Some(backtrace)) = len_source_backtrace {
+        // ...then infer other field as source field...
+        let source = (backtrace + 1) % 2;
+        // ...unless it was explicitly marked as non-source.
+        if parsed_fields.data.infos[source].info.source != Some(false) {
+            return Some(source);
+        }
+    }
```

Comment:
```suggestion
    // If we have exactly two fields...
    if state.fields.len() != 2 {
        return None;
    }
    // ... and no source field was specified/inferred...
    if parsed_fields.source.is_some() {
        return None;
    }

    // ... and one of the fields was specified/inferred as backtrace field,
    if let Some(backtrace) = parsed_fields.backtrace {
        // ...then infer other field as source field...
        let source = (backtrace + 1) % 2;
        // ...unless it was explicitly marked as non-source.
        if parsed_fields.data.infos[source].info.source != Some(false) {
            return Some(source);
        }
    }
```

This was sort of what I meant, but I agree that indeed it doesn't make it much more readable. The comments were most important.

Reply:
@JelteF Fixed.